### DISCUSSION TITLE
Help with {{selection}}

### DISCUSSION THREAD

**Discussion by tinoargentino** [Taitava/obsidian-shellcommands#333] [Q&A] [answered]
I’m trying to make a keyboard shortcut to auto launch specific commands that I store in my notes, so i've been playing with {{selection}} with the idea of selecting my saved command and hitting the keyboard shortcut

I created for instance one that works:

`open {{selection}}` so I select a path, launch it, and it works

However, I want to make it more general, so I tried with simple making the command:

`{{selection}}` 

with the idea of having it execute whatever the selected command is, but fails when i try it with the command above, I suspect it inserts weird spacing into the command but not sure

Am I missing something?

Is there a better way to do this?

**Comment by Taitava** [comment, marked as answer]
Hi, thank you for your question! 🙂

Yea, I guess I know the reason why it's not working, even without seeing the actual error message you are getting. I'll help you achieve what you are trying to do, but I'll first tell a bit of background for your problem. `{{Variables}}` are typically used as _arguments_ for shell commands, not (usually) as _actual commands_. [Special characters (including whitespace) in values coming from `{{variables}}` are escaped](https://publish.obsidian.md/shellcommands/Variables/Escaping+special+characters+in+variable+values) in order to prevent the values from accidentally breaking out of argument context, which could cause problems.

As your aim is not to use `{{selection}}` as an _argument_, you can [disable escaping](https://publish.obsidian.md/shellcommands/Variables/Escaping+special+characters+in+variable+values#Prevent+escaping+-+but+only+after+a+deep+consideration) by preceding the variable's name with an exclamation mark `!`. I.e. your shell command should be: `{{!selection}}`

Please let me know if you still encounter problems with this. 🙂 

# An alternative for `{{selection}}`

The above solution should work ok, but in case you want to avoid the need to highlight text in your notes before you can execute shell commands, one thing that came to my mind, is that you might try [Shell commands URI](https://publish.obsidian.md/shellcommands/Basic+usage/Shell+commands+URI). Shell commands URI allows you to create links that contain an executable shell command's id, and value(s) for [custom variable(s)](https://publish.obsidian.md/shellcommands/Variables/Custom+variables). I'll write instructions below so it's easier for you to see the idea:

1. [Create a custom variable](https://publish.obsidian.md/shellcommands/Variables/Custom+variables#How+to+create+custom+variables) and name it for example `{{_command}}`. You can give it another name, too. This variable will hold a shell command that will be executed.
2. Create a shell command whose content is simply just this: `{{!_commmand}}`
    - Note that escaping is disabled with the exclamation mark `!` here, too.
3. Generate a URL for the shell command. In the settings, next to the shell command you just created is an icon titled with _Copy this shell command's Obsidian uri to the clipboard_.
    - The link you just created looks similar to this one: `obsidian://shell-commands/?vault=MyVault&execute=amhtt94xn4&_command=` (`amhtt94xn4` is a shell command id, which will be different for you).
4. Append whatever shell command to the end of the link (after `_command=`).
    - An annoyance with this is that you need to URL encode the value of `_command`. There is a simple online URL encoder: https://www.urlencoder.io/
    - Example shell command that simply outputs `Hello world`: `obsidian://shell-commands/?vault=MyVault&execute=amhtt94xn4&_command=echo%20Hello%20world` 
5. You can create this kind of links as many as you want. In the settings, you'll only have one shell command _setting item_ defined, but as it uses a custom variable, it can execute different shell command contents. The contents will reside in your notes as links.

Please let me know whether this suits your use-case, or if you'd like to have something different. 🙂 

Edit: I'll move this discussion from _Debugging and testing_ to _Q & A_.

**Comment by tinoargentino** [comment]
thanks! that totally worked!

One trick is to use the Keyshots extension with the command "Encode selection" to quickly turn the command into url encoding

I’m sure there has to be a way to use your extension with selection to:

- grab selection
- encode it
- turn it into a markdown hyperlink

but i haven't figured it out yet. I will give it a thought. but if you have a proposal i'd love to hear it!

**Comment by Taitava** [comment]
Nice to know that it works 🙂

Yea, it's probably possible to come up with a shell command that takes `{{selection}}`, does url encoding for it, and outputs a link. I did some quick search and found that perhaps you could use `jq` command: https://stackoverflow.com/a/34407620/2754026 . I haven't tested that myself at all, so I don't even know if the command needs to be installed separately, or does it come preinstalled with macOS. (I assume you're using macOS as you mentioned the `open` command).

---

However, yesterday I came to think about creating an Obsidian command (i.e. not shell command) for creating execution links. It would open up a modal where users could enter values for customs variable(s). Those values would be automatically url encoded. But I don't have a schedule for it.
